src/core/service.c

```diff
@@ -2441,6 +2490,109 @@ static int service_deserialize_item(Unit *u, const char *key, const char *value,
                         s->watchdog_override_enable = true;
                         s->watchdog_override_usec = watchdog_override_usec;
                 }
+        } else if (streq(key, "main-command") || streq(key, "control-command")) {
+                bool control, found = false;
+                ServiceExecCommand id;
+                ExecCommand *command = NULL;
+                const char *word, *parser_state;
+                _cleanup_free_ char *args = NULL, *path = NULL;
+                _cleanup_strv_free_ char **argv = NULL;
+                size_t l;
+                int index, i;
+
+                enum ExecCommandState {
+                        STATE_EXEC_COMMAND_TYPE,
+                        STATE_EXEC_COMMAND_ID,
+                        STATE_EXEC_COMMAND_PATH,
+                        STATE_EXEC_COMMAND_ARGS,
+                        _STATE_EXEC_COMMAND_MAX,
+                        _STATE_EXEC_COMMAND_INVALID = -1,
+                } state;
+
+                control = streq(key, "control-command");
+
+                state = STATE_EXEC_COMMAND_TYPE;
+
+                FOREACH_WORD(word, l, value, parser_state) {
+                        _cleanup_free_ char *temp = NULL;
+
+                        temp = strndup(word, l);
+                        if (!temp)
+                                return log_oom();
+
+                        switch (state) {
+                        case STATE_EXEC_COMMAND_TYPE:
+                                id = service_exec_command_from_string(temp);
+                                state = STATE_EXEC_COMMAND_ID;
+                                break;
+                        case STATE_EXEC_COMMAND_ID:
+                                r = sscanf(temp, "%d", &index);
+                                if (r != 1) {
+                                        log_unit_debug(u, "Failed to deserialize index of currently executed command, assuming 0");
+                                        index = 0;
+                                }
+                                state = STATE_EXEC_COMMAND_PATH;
+                                break;
+                        case STATE_EXEC_COMMAND_PATH:
+                                path = temp;
+                                temp = NULL;
+                                state = STATE_EXEC_COMMAND_ARGS;
+
+                                assert(path_is_absolute(path));
```

Comment:
during deserialization we try to be defensive when it comes to data we read, after all we might do live updates from an older systemd version (or even live downgrades from a newer one!), and hence parse failures should not be fatal, and not even something we log about at a level higher than debug... An assert() is definitely too hard.